File: utils/reverse_policy.py
"""Tomislav-RetCtx: configure SDK reverse routing through collector discovery."""

import math

POLICIES = ("ttl", "probability", "inverse_depth", "depth_linear")


def validate(policy=None, probability=None):
    if policy is not None and policy not in POLICIES:
        raise ValueError("--reverse-policy must be " + "|".join(POLICIES))
    if policy == "probability":
        if (type(probability) not in (float, int) or not 0 <= probability <= 1
                or not math.isfinite(probability)):
            raise ValueError("--reverse-policy probability requires --reverse-probability in [0,1]")
    elif probability is not None:
        raise ValueError("--reverse-probability requires --reverse-policy probability")


def configure(config_map, policy=None, probability=None):
    validate(policy, probability)
    if policy is not None:
        config_map["reverse_policy"] = policy
        config_map.pop("reverse_probability", None)
        if policy == "probability":
            config_map["reverse_probability"] = probability
    else:
        # Preserve custom YAML when no override was supplied, validating it
        # with the same policy/p pairing as the collector and SDK.
        existing = config_map.get("reverse_policy", "ttl")
        if existing not in POLICIES:
            raise ValueError("invalid config_map.reverse_policy")
        if existing != "probability" and "reverse_probability" in config_map:
            raise ValueError("config_map.reverse_probability requires probability policy")
        value = config_map.get("reverse_probability")
        if isinstance(value, str):
            try:
                value = float(value)
            except ValueError:
                raise ValueError("invalid config_map.reverse_probability") from None
        validate(existing, value)

File: runtime/plugins/otelcol/reverse_policy.go
package otelcol

import (
	"encoding/json"
	"errors"
	"fmt"
	"math"
	"strconv"

	"github.com/blueprint-uservices/blueprint/runtime/core/backend"
)

var errInvalidReversePolicy = errors.New("invalid reverse checkpoint policy")

const (
	reversePolicyTTL          = "ttl"
	reversePolicyProbability  = "probability"
	reversePolicyInverseDepth = "inverse_depth"
	reversePolicyDepthLinear  = "depth_linear"
)

// Tomislav-RetCtx: collector-discovered reverse routing, immutable before spans
// start. Zero value keeps TTL behavior; probability policies need no new carrier
// metadata because every truss already contains its leaf's absolute depth.
type reversePolicy struct {
	mode        string
	probability float64
}

func parseReversePolicy(config map[string]interface{}) (reversePolicy, error) {
	policy := reversePolicy{mode: reversePolicyTTL}
	if value, present := config["reverse_policy"]; present {
		mode, ok := value.(string)
		if !ok {
			return reversePolicy{}, fmt.Errorf("%w: reverse_policy must be a string", errInvalidReversePolicy)
		}
		policy.mode = mode
	}
	value, hasProbability := config["reverse_probability"]
	switch policy.mode {
	case reversePolicyTTL, reversePolicyInverseDepth, reversePolicyDepthLinear:
		if hasProbability {
			return reversePolicy{}, fmt.Errorf("%w: reverse_probability requires reverse_policy: probability", errInvalidReversePolicy)
		}
	case reversePolicyProbability:
		probability, ok := reverseProbabilityValue(value)
		if !hasProbability || !ok {
			return reversePolicy{}, fmt.Errorf("%w: probability policy requires finite reverse_probability in [0,1]", errInvalidReversePolicy)
		}
		policy.probability = probability
	default:
		return reversePolicy{}, fmt.Errorf("%w: reverse_policy must be ttl, probability, inverse_depth, or depth_linear", errInvalidReversePolicy)
	}
	return policy, nil
}

func reverseProbabilityValue(value interface{}) (float64, bool) {
	var probability float64
	var err error
	switch v := value.(type) {
	case float64:
		probability = v
	case int:
		probability = float64(v)
	case int64:
		probability = float64(v)
	case json.Number:
		probability, err = v.Float64()
	case string:
		probability, err = strconv.ParseFloat(v, 64)
	default:
		return 0, false
	}
	return probability, err == nil && !math.IsNaN(probability) && probability >= 0 && probability <= 1
}

func (p reversePolicy) probabilistic() bool {
	return p.mode != "" && p.mode != reversePolicyTTL
}

// Tomislav-RetCtx: n is this truss's leaf depth, not the deepest sibling's depth;
// d is the current receiver depth. Linear probabilities sum to one across the
// n potential receivers (d=0..n-1), just like 1/n, but favor deeper receivers.
// These are conditional per-node probabilities, not final emission weights.
func (p reversePolicy) probabilityAt(d, n uint64) float64 {
	if n == 0 || d >= n {
		return 0 // Invalid ancestry cannot supply a depth-based decision.
	}
	switch p.mode {
	case reversePolicyProbability:
		return p.probability
	case reversePolicyInverseDepth:
		return 1 / float64(n)
	case reversePolicyDepthLinear:
		// Convert before adding/multiplying to avoid uint64 overflow.
		return (2 * ((float64(d) + 1) / float64(n))) / (float64(n) + 1)
	default:
		return 0
	}
}

func (p reversePolicy) route(retCtx string, originalCheckpoint bool, depth uint64, draw func() float64) (string, string) {
	if !p.probabilistic() {
		return backend.RouteRetCtx(retCtx, originalCheckpoint)
	}
	return backend.RouteRetCtxWithDecision(retCtx, originalCheckpoint, func(cp backend.ReturnedCheckpoint) bool {
		probability := p.probabilityAt(depth, cp.Depth)
		return probability >= 1 || (probability > 0 && draw() < probability)
	})
}
